narwhal/worker/src/batch_fetcher.rs

```diff
@@ -0,0 +1,530 @@
+    ) -> anyhow::Result<HashMap<BatchDigest, Batch>> {
+        let mut verified_batches = HashMap::new();
+        let mut digests_to_fetch = digests.clone();
+
+        // Continue to bulk request from remote worker until no remaining
+        // digests are available.
+        loop {
+            if digests_to_fetch.is_empty() {
+                break;
+            }
+            let RequestBatchesResponse {
+                batches,
+                is_size_limit_reached,
+            } = self
+                .network
+                .request_batches(
+                    digests_to_fetch.clone().into_iter().collect(),
+                    worker.clone(),
+                    timeout,
+                )
+                .await?;
+            // Use this flag to determine if the worker returned a valid digest
+            // to protect against an adversarial worker returning size limit reached
+            // and no batches leading us to retry this worker forever or until
+            // we get batches from other workers.
+            let mut is_digest_received = false;
+            for batch in batches {
+                let batch_digest = batch.digest();
+                if !digests_to_fetch.contains(&batch_digest) {
+                    bail!("[Protocol violation] Worker {worker} returned batch with digest {batch_digest} which is not part of the requested digests: {digests:?}");
+                } else {
+                    is_digest_received = true;
+                    verified_batches.insert(batch_digest, batch);
+                    digests_to_fetch.remove(&batch_digest);
+                }
+            }
+            if !is_size_limit_reached || !is_digest_received {
+                break;
+            }
+        }
+
+        Ok(verified_batches)
+    }
+}
+
+// todo - make it generic so that other can reuse
+struct PendingGuard<'a> {
+    metric: &'a IntGauge,
+}
+
+impl<'a> PendingGuard<'a> {
+    pub fn make_inc(metric: &'a IntGauge) -> Self {
+        metric.inc();
+        Self { metric }
+    }
+}
+
+impl<'a> Drop for PendingGuard<'a> {
+    fn drop(&mut self) {
+        self.metric.dec()
+    }
+}
+
+// Trait for unit tests
+#[async_trait]
+pub trait SubscriberNetwork: Send + Sync {
+    async fn request_batches(
+        &self,
+        batch_digests: Vec<BatchDigest>,
+        worker: NetworkPublicKey,
+        timeout: Duration,
+    ) -> anyhow::Result<RequestBatchesResponse>;
+}
+
+struct SubscriberNetworkImpl {
+    network: anemo::Network,
+}
+
+#[async_trait]
+impl SubscriberNetwork for SubscriberNetworkImpl {
+    async fn request_batches(
+        &self,
+        batch_digests: Vec<BatchDigest>,
+        worker: NetworkPublicKey,
+        timeout: Duration,
+    ) -> anyhow::Result<RequestBatchesResponse> {
+        let request =
+            anemo::Request::new(RequestBatchesRequest { batch_digests }).with_timeout(timeout);
+        self.network.request_batches(worker, request).await
+    }
+}
+
+#[cfg(test)]
```

Comment:
I believe most tests are migrated, except the multi-worker test which will be done in Subscriber.